### ISSUE TITLE
Changing Mirage's default linkage data behavior in 0.4 to better conform to JSON:API's semantics

### ISSUE THREAD

**Issue by samselikoff** [samselikoff/ember-cli-mirage#1475]
I wanted to write a quick post detailing a change that happened in Mirage's 0.4.0 release that I originally didn't signal as a breaking change, but that affected more apps than I realized.

I have since added the following note about it to the [0.3 -> 0.4 upgrade guide]({{ site.baseurl }}{% link docs/v0.4.x/upgrading.md %})
, and I'm reposting it here for more visibility. I think it's worth understanding, as the issue sheds light on the semantics around some of JSON:API's design decisions.

Apologies if you upgraded to 0.4 and had to debug this yourself!

---

There is one primary change in 0.4 that could break your 0.3 app.

In 0.3.x, Mirage's JSONAPISerializer included all related foreign keys whenever serializing a model or collection, even if those relationships were not being `included` in the payload.

This actually goes against JSON:API's design. Foreign keys in the payload are known as [Resource Linkage](http://jsonapi.org/format/#document-resource-object-linkage) and are intended to be used by API clients to link together all resources in a JSON:API compound document. In fact, most server-side JSON:API libraries do not automatically serialize all related foreign keys, and only return linkage data for related resources when they are being included in the current document.

By including linkage data for every relationship in 0.3, it was easy to develop Ember apps that would work with Mirage but would behave differently when hooked up to a standard JSON:API server. Since Mirage always included linkage data, an Ember app might automatically be able to fetch related resources using the ids from that linkage data plus its knowledge about the API. For example, if a `post` came back like this:

```js
// GET /posts/1
{
  data: {
    type: 'posts',
    id: '1',
    attributes: { ... },
    relationships: {
      author: {
        data: {
          type: 'users',
          id: '1'
        }
      }
    }
  }
}
```

and you forgot to `?include=author` in your GET request, Ember Data would potentially use the `user:1` foreign key and lazily fetch the `author` by making a request to `GET /authors/1`. This is problematic because

1. This is not how foreign keys are intended to be used
2. It'd be better to see no data and fix the problem by going back up to your data-loading code and add `?include=author` to your GET request, or
3. If you do want your interface to lazily load the author, use resource `links` instead of the resource linkage `data`:

```js
// GET /posts/1
{
  data: {
    type: 'posts',
    id: '1',
    attributes: { ... },
    relationships: {
      author: {
        links: {
          related: '/api/users/1'
        }
      }
    }
  }
}
```

Resource links can be defined on Mirage serializers using the [links](http://www.ember-cli-mirage.com/docs/v0.3.x/serializers/#linksmodel) method (though `including` is likely the far more simpler and common approach to fetching related data).

So, Mirage 0.4 changed this behavior and by default, the JSONAPISerializer only includes linkage data for relationships that are being included in the current payload (i.e. within the same compound document).

This behavior is configurable via the `alwaysIncludeLinkageData` key on your JSONAPISerializers. It is set to `false` by default, but if you want to opt-in to 0.3 behavior and always include linkage data, set it to `true`:

```js
// mirage/serializers/application.js
import { JSONAPISerializer } from 'ember-cli-mirage';

export default JSONAPISerializer.extend({
  alwaysIncludeLinkageData: true
});
```

If you do this, I would recommend looking closely at how your real server behaves when serializing resources' relationships and whether it uses resource `links` or resource linkage `data`, and to update your Mirage code accordingly to give you the most faithful representation of your server.

**Comment by samselikoff** [comment]
This is a comment.

**Comment by samselikoff** [comment]
Another comment

**Comment by Herriau** [comment]
@samselikoff I wouldn't say this better conforms with the JSONAPI spec. I believe this may actually be a slight step backward.

While the JSONAPI spec itself is fairly ambiguous about whether the server may omit some (or all) relationships from a resource object's [relationships object](https://jsonapi.org/format/#document-resource-object-relationships) when the client hasn't requested any particular sparse fieldset (see https://github.com/json-api/json-api/issues/1388), this change, at a minimum, goes against JSONAPI's recommendations:

> The base specification is agnostic about including links with a resource response. However, it is recommended that the following links be included within response documents:
> 
> - **Top-level links** like a self-link (for the whole response) as well as relative pagination links (if appropriate).
> - **Resource-level links** like a self-link for each resource (which differs from the top-level, if the resource is part of a collection).
> - **Relationship links** for all available relationships of a resource.

(From the following recommendation: https://jsonapi.org/recommendations/#including-links)

I wish Mirage would better align with JSONAPI serializer libs throughout which a common behavior is to:
- Have all attributes and relationships be represented in the [resource object](https://jsonapi.org/format/#document-resource-objects)'s [attributes object](https://jsonapi.org/format/#document-resource-object-attributes) and [relationships object](https://jsonapi.org/format/#document-resource-object-relationships) respectively, unless otherwise requested by a sparse fieldset.
- Have all relationship objects contain a [links object](https://jsonapi.org/format/#document-links) by default (unless otherwise specified by what these libraries commonly refer to as "default includes", in which case [resource linkage](https://jsonapi.org/format/#document-resource-object-linkage) would be included in addition to the [links object](https://jsonapi.org/format/#document-links)).
- When the client supplies an explicit list of includes, then the relationships corresponding to these includes will be "expanded" in that their relationship object will contain [resource linkage](https://jsonapi.org/format/#document-resource-object-linkage) in addition to their [links object](https://jsonapi.org/format/#document-links) while relationships that were not explicitly requested will still be represented and their relationship object will contain a [links object](https://jsonapi.org/format/#document-links).

To summarize, if I have a model defined as:

```javascript
// models/my-resource.js
export default Model.extend({
    foo: attr(...),
    bar: attr(...),
    baz: belongsTo(...),
    qux: hasMany(...),
});
```

I would expect a request to `GET /my-resources/123?include=baz` to generate the following response:

```json
{
    "data": {
        "type": "my-resources",
        "id": 123,
        "attributes": {
            "foo": ...,
            "bar": ...
        },
        "relationships": {
            "baz": {
                "links": { ... },
                "data": { ... }
            },
            "qux": {
                "links": { ... }
            }
        }
    },
    "included": [
        ...
    ]
}
```

**Comment by samselikoff** [comment]
Looking at your example, I think the only thing Mirage doesn't do out of the box is define `links` on the `relationships` hash, and that's because Mirage has no higher-level routing concept baked into it. (You always only define single routes at a time.)

However, there is [a `links` API on Serializers](http://www.ember-cli-mirage.com/docs/api/modules/ember-cli-mirage/serializer~Serializer#links) which can be used for this purpose:

```js
import { JSONAPISerializer } from 'ember-cli-mirage';

export default JSONAPISerializer.extend({

  links(myResource) {
    return {
      baz: {
        links: {
          self: `/my-resources/${myResource.id}/relationships/baz`,
          related: `/my-resources/${myResource.id}/baz`  
        }
      }
      qux: {
        links: {
          self: `/my-resources/${myResource.id}/relationships/qux`,
          related: `/my-resources/${myResource.id}/qux`  
        }
      }  
    };
  }

});
```

Obviously this would be cumbersome to implement for every relationship on every resource. In the past when this is important to me, I've added a bit of dynamic code to solve this. Something like the following (pseudocode):

```js
import { JSONAPISerializer } from 'ember-cli-mirage';

export default JSONAPISerializer.extend({

  links(resource) {
    let links = {};
    
    resource.associationKeys.forEach(key => {
      links[key] = {
        links: {
          self: `/${pluralize(resource.modelName)}/${resource.id}/relationships/${key}`,
          related: `/${pluralize(resource.modelName)}/${resource.id}/${key}`  
        }
      }
    });

    return links;
  }

});
```

This would require the related endpoints be defined, which could also be done somewhat dynamically:

```js
// mirage/config.js
this.get('my-resources/:id/relationships/:relationship_name', (schema, request) => {
  let myResource = schema.myResources.find(request.params.id);
  
  return myResource[request.params.relationship_name];
});
```

The reason we haven't baked any of this in yet is because the `links` strategy for backends differ - there is nothing in the spec that indicates what the link relationship should look like. However, we absolutely could and should provide some routing abstraction that makes this much easier.

The main reason this post claimed that the change related to linkage _data_ (as opposed to `links`) better conformed to the spec, is because linkage data is intended to link resources within a single document, rather than be used to fetch remote resources by ID.

**Comment by Herriau** [comment]
@samselikoff Thank you for the quick reply! That makes sense. I definitely think Mirage should ship with some of the code you included in your comment as this is something people are likely to have to reinvent over and over again. 

Perhaps the JSONAPISerializer could expose a hook that given a model instance and a relationship key would return the `self` and/or `related` link(s) with a reasonable default implementation.
